### PR TITLE
Unit testing and behavior tests

### PR DESCRIPTION
**Pull request by miniche**
The CentralReport project is growing, it's time to implement tests!
Tests allow to ensure the integrity of the application is kept between versions and to ensure that all functionalities works well all the time.

CentralReport have two main parts:
- The bash part (installer/uninstaller and other related tools)
- The Python part (the application)

This pull request creates the basis of unit and system tests. There is a lot of work with the amount of functions and methods. All tests will be implemented while CentralReport is developed.

The following methods are ready:
- [x] Tests for the installer and the uninstaller (system tests)
- [x] Unit tests for the "utils" Python package
- [x] Tests with [Travis-CI](https://travis-ci.org/CentralReport/CentralReport), supporting the tests on multiple Python versions
- [x] Tests with [Vagrant](http://www.vagrantup.com/)
## Start tests

All tests can be run with the `run_tests.sh` script, in the "tests" directory. This script accepts the following options:

``` bash
-a Performs all tests, including the Vagrant tests
-p Only performs the Python unit tests
-s Only performs the system tests (installer/uninstaller)
-v Only use Vagrant to test CentralReport on virtual machines (in this case, Python unit tests and system tests are performed on each VM)
```

tests
## Vagrant

Vagrant is a powerful software to test your project on many virtual machines, with the help of VirtualBox or VMware.
To start, I have set up three virtual machines:
- Debian 6.0.5 x64
- CentOS 6.4 x64
- Ubuntu Server 12.04 x64

They are hosted at http://static.centralreport.net/vagrant/.
If you want to test this feature, keep in mind that the size of each virtual machine is approximately 300 Mb.
It requires the lastest Vagrant version and VirtualBox >= 4.2
## In the future

There are lots of tests to write now. To avoid a long time wasting, they will be written while new features are added.
TODO:
- Unit tests for all bash functions (with shunit2)
- More tests for bash methods (install/uninstall/manager/CLI/...)
- More unit tests for Python
- System tests for Python
## Remains to be done in this pull request
- [ ] Update the NEWS file

### CODE REVIEW

**Comment by miniche** [comment]
Guys,

This pull request has been opened six months ago... It's time to merge it in develop!
There is a lot of work to write all the tests. I think it's better to use the tests already implemented and write some new tests each week while new features are created.

I have updated the Pull Request description. 
So... Ready for your review :)

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  config.vm.define "cr_centos_6_4_64" do |cr_centos_6_4_64|
+    cr_centos_6_4_64.vm.box = "cr_centos_6_4_64"
+    cr_centos_6_4_64.vm.box_url = "http://static.centralreport.net/vagrant/centos_6_4_64.box"
+  end
+
+  config.vm.define "cr_ubuntu_server_12_04_64" do |cr_ubuntu_server_12_04_64|
+    cr_ubuntu_server_12_04_64.vm.box = "cr_ubuntu_server_12_04_64"
+    cr_ubuntu_server_12_04_64.vm.box_url = "http://static.centralreport.net/vagrant/ubuntu_server_12_04_x64.box"
+  end
+end
```
Comment (Ninir) [inline Vagrantfile:23]: :+1: This is just awesome for tests :)

**Comment by Ninir** [comment]
Will check it tomorrow using vagrant (need to dl 3 VMs :P)

**Comment by Ninir** [comment]
A few things:
- [x] Is it possible to set success messages in green rather than yellow?
- [x] Running the tests asked for my password twice during the process. Isn't it possible to store the password, or set a "sudo-like session"?
- [x] When running run_tests.sh without options, it seems that no/only a few tests are run. Isn't it possible to set the "a" option by default?

**Comment by miniche** [comment]
@Ninir:
- Yes, the green color is more appropriate than the yellow color. But the default Unix green is... hum... very hard to define :P Look at the PR #29 (Removed green lightbox). I can make test with green and black font if you want.
- This behavior occurs only on Mac OS. On other systems, tests are performed as root and no need a "sudo" session. In fact, on OS X, the "sudo session" is created when the installer/uninstaller are starting and they are destroyed at the end of the processus for security purposes. I can add an option to the installer ("-k" to "keep the sudo session alive" for example) to not destroy this sudo session at the end of the installer when tests are performed. Thus the password will be asked only when the tests begin.
- Yeah, that is not intended. We can to set the "a" option by default, even if it could be very annoying: the tests with Vagrant take several minutes. I can add a test if Vagrant is available on the host to avoid errors on virtual machines.

**Comment by miniche** [comment]
@Ninir:
The "sudo session" has been added with the commits a45cb5c and dd4d0b0. The password is only asked to the user on OS X if the "system tests" will be performed.

**Comment by miniche** [comment]
Green box: the return! I have added the green color for bash outputs (printLightBox function - commit d9c0e6).

A screenshot of this color:
green color

@Ninir, is it good for you? :)

**Comment by miniche** [comment]
The default option for the "run_tests.sh" script is now "-a". If no option is supplied, all tests will be performed (commit a4e92c3).
In this case, the tests on virtual machines with Vagrant may take several minutes and may annoying if the users had just forgotten an argument (like -p or -s). If no option is supplied, a message is displayed to warn the user, and he may cancel the test for 10 seconds. After 10 seconds, all tests are performed.

**Comment by Ninir** [comment]
Awesome! Checked by my side. Ready to go, isn't it? :smile:

**Comment by miniche** [comment]
Yeah!
I must perform more tests on some operating systems before merging, and update the NEWS file too.

The project will be paused during the next two weeks while I move and find an Internet connection :)

**Comment by miniche** [comment]
Back to work!

I have performed more tests today, and I must fix some return values. I will do it this week.

**Comment by miniche** [comment]
Good.
I have fixed some bugs with Vagrant. @Ninir, can you perform some tests?

It's good for me, ready to merge! Waiting for your approval.

**Comment by Ninir** [comment]
It lacks some style and space in CLI, otherwise it's just awesome, congratz :)

**Comment by miniche** [comment]
Alright, it can be improved in the future.
I will merge this PR to start tests on other branches :)

**Pull request merged**